contracts/usul/BaseTokenVoting.sol

```diff
@@ -0,0 +1,172 @@
+    function updateTimelockPeriod(
+        uint256 _newTimelockPeriod
+    ) external onlyOwner {
+        _updateTimelockPeriod(_newTimelockPeriod);
+    }
+
+    /// @notice Updates the voting time period
+    /// @param _newVotingPeriod The voting time period in seconds
+    function _updateVotingPeriod(uint256 _newVotingPeriod) internal {
+        votingPeriod = _newVotingPeriod;
+
+        emit VotingPeriodUpdated(_newVotingPeriod);
+    }
+
+    /// @notice Updates the timelock period - time between queuing and when a proposal can be executed
+    /// @param _newTimelockPeriod The new timelock period in seconds
+    function _updateTimelockPeriod(uint256 _newTimelockPeriod) internal {
+        timelockPeriod = _newTimelockPeriod;
+
+        emit TimelockPeriodUpdated(_newTimelockPeriod);
+    }
+
+    /// @notice Returns true if an account has voted on the specified proposal
+    /// @param _proposalId The ID of the proposal to check
+    /// @param _account The account address to check
+    /// @return bool Returns true if the account has already voted on the proposal
+    function hasVoted(
+        uint256 _proposalId,
+        address _account
+    ) public view returns (bool) {
+        return proposals[_proposalId].hasVoted[_account];
+    }
+
+    /// @notice Function for counting a vote for a proposal, can only be called internally
+    /// @param _proposalId The ID of the proposal
+    /// @param _voter The address of the account casting the vote
+    /// @param _support Indicates vote support, which can be "No", "Yes", or "Abstain"
+    /// @param _weight The amount of voting weight cast
+    function _vote(
+        uint256 _proposalId,
+        address _voter,
+        uint8 _support,
+        uint256 _weight
+    ) internal {
+        require(
+            proposals[_proposalId].deadline != 0,
+            "Proposal has not been submitted yet"
+        );
+        require(
+            block.timestamp <= proposals[_proposalId].deadline,
+            "Voting period has passed"
+        );
+        require(
+            !proposals[_proposalId].hasVoted[_voter],
+            "Voter has already voted"
+        );
+
+        proposals[_proposalId].hasVoted[_voter] = true;
+
+        if (_support == uint8(VoteType.Against)) {
+            proposals[_proposalId].noVotes += _weight;
+        } else if (_support == uint8(VoteType.For)) {
+            proposals[_proposalId].yesVotes += _weight;
+        } else if (_support == uint8(VoteType.Abstain)) {
+            proposals[_proposalId].abstainVotes += _weight;
+        } else {
+            revert("Invalid value for enum VoteType");
+        }
+
+        emit Voted(_voter, _proposalId, _support, _weight);
+    }
+
+    /// @notice Notifies the strategy of a new proposal, only callable by the Usul Module
+    /// @param _data Encoded proposal data, in this implementation only includes proposal ID
+    function receiveProposal(
```

Comment:
I wonder if there's a better naming than "receiveProposal" for this.  To me it seems similar to the "getProposal" function name, but does something else entirely

Reply:
Yeah agreed, don't really like the naming on this either. How about `initializeProposal`?

Reply:
yeah that sounds good, thank you sir

Reply:
Pushed this one up